### ISSUE TITLE
[Bug] Conditional validate prop

### ISSUE THREAD

**Issue by sethomas** [final-form/react-final-form#372]
### Are you submitting a **bug report** or a **feature request**?

Bug

### What is the current behavior?

When conditionally passing in a validate function, it does not register until a form change occurs.

### What is the expected behavior?

When conditionally passing in a validate function, it will trigger the error.

### Sandbox Link

https://codesandbox.io/s/xrko2m3wpw

### What's your environment?

"node": ">=8.11.1 <9.0.0",
"npm": ">=5.6.0"

"final-form": "^4.10.0",
"react-final-form": "^3.6.7",
"react-final-form-listeners": "^1.0.1",

### Other information

 - Press Submit
 - Notice first name validation error (awesome)
 - Check the "Require Last Name" checkbox

Expected: would see the error for last name required
Actual: Does not show validation error unless another field changes

**Comment by sethomas** [comment]
For a janky fix, I am using `form.change` twice, one time to add a space to the conditional validate field, and another to remove the space.

**Comment by emartini** [comment]
You can use a cleaner hack in the meanwhile:

```js
const required = value => (value ? undefined : "Required");
const requiredLastName = (value, allValues) =>
  allValues.requireLastName ? required(value) : undefined;
```

```jsx
 <Field name="lastName" validate={requiredLastName}>
```

https://codesandbox.io/s/km5k1jy6oo

**Comment by bogdansoare** [comment]
also having this issue

**Comment by davidroeca** [comment]
I've encountered this bug as well. My only non-crazy workaround is to remove field-level validation and switch to submit validation for the relevant fields.

**Comment by Jony-Y** [comment]
Any upcoming fix for this?

**Comment by KajetanDobczyk** [comment]
+1 from me. I lost at a lot of time because of this, as i was using formProps.values in conditional validation, and they weren't updated. I had to create conditional validator for each field instead, using allValues within.

**Comment by werol** [comment]
I was struggling with the same problem and the solution is described in the documentation even with the codesandbox attached.

> ⚠️ IMPORTANT ⚠️ – By default, in order to allow inline fat-arrow validation functions, the field will not rerender if you change your validation function to an alternate function that has a different behavior. If you need your field to rerender with a new validation function, you will need to update another prop on the Field, such as key. See the following sandbox for an example:

https://codesandbox.io/s/changing-field-level-validators-zc8ei?fontsize=14

**Comment by erikras** [comment]
Do people understand the reasoning presented in the docs, so kindly shared here by @werol? It's a trade-off that I think would annoy many more people if it were the other way around (i.e. having to reregister a field _on every render_ if a `=>` function were provided to `validate`).

If you can think of a better way to explain it, I'm all ears. Not sure how it could be [more prominent in the docs](https://final-form.org/docs/react-final-form/types/FieldProps#validate).

**Comment by davidroeca** [comment]
Agreed that this makes it easier for beginners to use, though an alternative would be to recommend `React.useCallback` for in-line arrow functions that don’t change, and to re-register on changed validate functions. Unclear on the performance trade-offs here, just a thought.

**Comment by ChrisChiasson** [comment]
Seconding what @davidroeca wrote. I've actually read the doc about the arrow function a few times in the past and still today I did not remember it, causing myself wasted time. The way I found this thread was actually by Googling: `react-final-form` `validate` `useCallback`... because that was the kind of function I switched to while trying to debug and it still wasn't updating. It sorta seems like over time, optimal use of React is changing, with hooks and the forthcoming concurrent mode.
